0000-crypto-construct-language.md

```diff
@@ -0,0 +1,761 @@
+identifier is specific to the application. In some cases it may be the file
+name of a file to open or it may be the identifier of a transaction in a
+blockchain application or it could any reference to a data object that makes
+sense in the application. The mode is an identifier representing read, write,
+and append. The result is a handle to the opened object that can be used by
+other data I/O commands.
+
+```
+/ h -- handle to the opened data storage object.
+/ s -- zero-indexed starting offset to begin the read.
+/ n -- number of bytes to read from the object.
+/ b -- binary read from the data object.
+READ ( h s n -- b h )
+```
+
+The `READ` opcode takes the handle to the opened data storage object and reads
+the number of bytes specified starting at the offset given. The result is the
+binary data read from the data object and then the handle to the open data
+object on top.
+
+```
+/ h -- handle to the opened data storage object.
+/ b -- binary data to write.
+WRITE ( h b -- h )
+```
+
+The `WRITE` opcode takes the handle and the data to write and writes it to the
+open data object. The result is the handle to the data object.
+
+```
+/ h -- handle to the opened data storage object.
+/ n -- number of bytes and direction to seek in.
+SEEK ( h n -- h )
+```
+
+The `SEEK` opcode seeks the specified number of bytes. If the number of bytes
+is negative, it seeks backwards towards the 0th index byte. If the number is
+positive, it seeks forward towards the last byte in the object. The result is
+the handle to the opened data object.
+
+```
+/ h -- handle to the opened data storage object.
+CLOSE ( h -- )
+```
+
+The `CLOSE` opcode closes the opened data storage object. It pops the handle
+from the stack and does not push anything to the stack.
+
+### Stack Control
+
+```
+/ a -- argument of any type.
+DUP ( a -- a a )
+```
+
+The `DUP` opcode pops the top item, duplicates it and pushes the original and
+its duplicate on the top of the stack.
+
+```
+/ a -- argument of any type.
+POP ( a -- )
+```
+
+The `POP` opcode pops the top item from the stack and forgets about it. This is
+used for throwing away the top of the stack.
+
+### Flow Control
+
+```
+/ b -- boolean argument.
+IF ( b -- ) ELSE THEN
+```
+
+The `IF-ELSE-THEN` opcode trio and the related `IF-THEN` opcode duo are used
+to do conditional branching in CCL scripts. The `IF` opcode pops the top of the
+stack and evaluates it as a boolean argument.
+
+In the case of `IF-ELSE-THEN` if the argument is true then the script between
+`IF` and `ELSE` is excuted followed by jumpting to the script after the `THEN`.
+If it is a false then the script between the `ELSE` and the `THEN` is executed.
+
+In the case of `IF-THEN` if the argument is true then script between the `IF`
+and `THEN` is executed. If it is false then flow jumps to the script after the
+`THEN` opcode.
+
+## Encoding Formats
+
+The first version of CCL supports the following encoding types:
```

Comment:
These are good choices (I'm not sure Base58 is needed, but it reduces confusion). I think you also need to specify the encoding a little bit more. For example, with Base64 I assume it is not with BEGIN and END lines--it is all one string.  But is it enclosed in quotes? Or just delimited with whitespace?

Reply:
I included more description on what I mean for each encoding type. I also included Base58 to show that this is intended to be extensible and application neutral. I also added language saying that Base58Check is a complex construct that can be stored in CCL as a script that unpacks and checks the hashes and decodes the Base58 pieces leaving the payload string in binary form on the stack.